src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+     * caller and no value for the element is recorded.
+     * <p>
+     * The returned list and its {@link List#subList(int, int) subList} views implement
+     * the {@link RandomAccess} interface.
+     * <p>
+     * The returned list is not {@link Serializable} and, as it is unmodifiable, does
+     * not implement the {@linkplain Collection##optional-operation optional operations}
+     * in the {@linkplain List} interface.
+     * <p>
+     * If the provided {@code mapper} recursively calls the returned list for the
+     * same index, an {@linkplain IllegalStateException} will be thrown.
+     *
+     * @param size   the size of the returned list
+     * @param mapper to invoke whenever an element is first accessed
+     *               (may return {@code null})
+     * @param <E>    the type of elements in the returned list
+     * @throws IllegalArgumentException if the provided {@code size} is negative.
+     */
+    static <E> List<E> list(int size,
+                            IntFunction<? extends E> mapper) {
+        if (size < 0) {
+            throw new IllegalArgumentException();
+        }
+        Objects.requireNonNull(mapper);
+        return SharedSecrets.getJavaUtilCollectionAccess().stableList(size, mapper);
+    }
+
+    /**
+     * {@return a new stable map with the provided {@code keys}}
+     * <p>
+     * The returned map is an {@linkplain Collection##unmodifiable unmodifiable} map whose
+     * keys are known at construction. The map's values are computed via the provided
+     * {@code mapper} when they are first accessed
+     * (e.g. via {@linkplain Map#get(Object) Map::get}).
+     * <p>
+     * The provided {@code mapper} function is guaranteed to be successfully invoked
+     * at most once per key, even in a multi-threaded environment. Competing
+     * threads accessing a value already under computation will block until an element
+     * is computed or an exception is thrown by the computing thread.
+     * <p>
+     * If the provided {@code mapper} throws an exception, it is relayed to the initial
+     * caller and no value associated with the provided key is recorded.
+     * <p>
+     * The returned map is not {@link Serializable} and, as it is unmodifiable, does
+     * not implement the {@linkplain Collection##optional-operations optional operations}
+     * in the {@linkplain Map} interface.
+     * <p>
+     * If the provided {@code mapper} recursively calls the returned map for
+     * the same key, an {@linkplain IllegalStateException} will be thrown.
+     *
+     * @param keys   the (non-null) keys in the returned map
+     * @param mapper to invoke whenever an associated value is first accessed
+     *               (may return {@code null})
+     * @param <K>    the type of keys maintained by the returned map
+     * @param <V>    the type of mapped values in the returned map
+     */
+    static <K, V> Map<K, V> map(Set<K> keys,
```

Comment:
Like `List`, a map can also generate other kinds of collections:
1. an entry set (`Map::entrySet`)
2. a key set (`Map::keySet`)
3. a value set (`Map::values`)

We should make it clear that calling (1) and (3) will result in materializing the values for the map -- we do not have "lazy" views for such sets.

Reply:
These constructs are also lazy actually which is cool. Operating on them will only materialize values on demand.

We should document that.